### ISSUE TITLE
control plane down/something changing on-disk file permisisons?

### ISSUE THREAD

**Issue by jansmets** [CrunchyData/postgres-operator#2838]
I am running 3 different clusters and all experienced the same issue. 
The control plane went down (eg, kubernetes upgrades) and that triggered a crash-recovery of the instance. The restart failed because the directory file permissions changed.

a) What could be changing the file permissions
b) Is there any resiliency during control-plane upgrades/downtime? "just keep on going"?

```
INFO: Lock owner: keycloak-keycloak-1-w7cg-0; I am keycloak-keycloak-1-w7cg-0
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f4880048c50>: Failed to establish a new connection: [Errno 111] Connection refused',)': /api/v1/namespaces/sr-sso/endpoints/keycloak-ha
ERROR: Request to server https://10.96.0.1:443 failed: MaxRetryError("HTTPSConnectionPool(host='10.96.0.1', port=443): Max retries exceeded with url: /api/v1/namespaces/sr-sso/endpoints/keycloak-ha (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f48800beef0>: Failed to establish a new connection: [Errno 111] Connection refused',))",)
...
INFO: doing crash recovery in a single user mode
ERROR: Crash recovery finished with code=1
INFO: stdout=
INFO: stderr=
[27668] FATAL: data directory "/pgdata/pg14" has invalid permissions
[27668] DETAIL: Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
```

**Comment by ibot3** [comment]
I experienced a similar problem. Both of my pg clusters had problems with the file permissions this night. I don't exactly know  why. But only the primary pods had the problems, the secondaries worked fine.
This caused some kind of a crash loop but no failover happened which would have prevented the outage.

```
2022-01-07 10:00:49,706 INFO: Lock owner: pg-test-pg-test-1-hqbv-0; I am pg-test-pg-test-1-hqbv-0
2022-01-07 10:00:49,715 INFO: starting as readonly because i had the session lock
2022-01-07 10:00:50.070 UTC [131816] FATAL:  data directory "/pgdata/pg13" has invalid permissions
2022-01-07 10:00:50.070 UTC [131816] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
2022-01-07 10:00:50,097 INFO: postmaster pid=131816
/tmp/postgres:5432 - no response
2022-01-07 10:00:50,118 WARNING: Postgresql is not running.
2022-01-07 10:00:50,119 INFO: Lock owner: pg-test-pg-test-1-hqbv-0; I am pg-test-pg-test-hqbv-0
2022-01-07 10:00:50,198 INFO: pg_controldata:
...

2022-01-07 10:00:50,222 INFO: doing crash recovery in a single user mode
2022-01-07 10:00:50,245 ERROR: Crash recovery finished with code=1
2022-01-07 10:00:50,245 INFO:  stdout=
2022-01-07 10:00:50,246 INFO:  stderr=2022-01-07 10:00:50.243 UTC [131822] FATAL:  data directory "/pgdata/pg13" has invalid permissions
2022-01-07 10:00:50.243 UTC [131822] DETAIL:  Permissions should be u=rwx (0700) or u=rwx,g=rx (0750).
```

**Comment by tolleiv** [comment]
Running into the same issues - I found that the operator had issues using the Kubernetes API alongside

Related logs from the pgo.

Once the issues occurs, PGO isn't able to reach the instance:
```
time="2022-01-31T09:33:11Z" level=error msg="Reconciler error" error="error dialing backend: dial tcp 10.61.4.65:10250: i/o timeout" file="internal/controller/postgrescluster/patroni.go:147" func="postgrescluster.(*Reconciler).reconcilePatroniDynamicConfiguration" name=db-instance namespace=application reconciler group=postgres-operator.crunchydata.com reconciler kind=PostgresCluster version=5.0.4-0
```

Some time later it tried to automatically mitigate the issues but fails:
```
time="2022-01-31T09:47:21Z" level=error msg="Reconciler error" error="error sending request: Post \"https://10.161.0.1:443/api/v1/namespaces/application/pods/db-instance-instance1-2swc-0/exec?command=patronictl&command=edit-config&command=--replace%3D-&command=--force&container=database&stderr=true&stdin=true&stdout=true\": read tcp 172.16.1.160:58150->10.161.0.1:443: read: connection reset by peer" file="internal/controller/postgrescluster/patroni.go:147" func="postgrescluster.(*Reconciler).reconcilePatroniDynamicConfiguration" name=db-instance namespace=application reconciler group=postgres-operator.crunchydata.com reconciler kind=PostgresCluster version=5.0.4-0
```

I haven't really figured out why this happens in our case but maybe this helps others to further debug this.

**Comment by budimanjojo** [comment]
just happen to me too. Manually changing the permission of directory `/pgdata/pg14` to 700 fixed it

**Comment by cbandy** [comment]
What version of Kubernetes, what storage provider, etc? What else is going on in the Kubernetes cluster when `FATAL:  data directory "…" has invalid permissions` is reported?

**Comment by budimanjojo** [comment]
Kubernetes version: k3s v1.23.4+k3s1
Storage is a Rook Ceph cluster
My situation is like the OP, after a node which has the statefulset running on is rebooted, the new pod is created and it failed to recover.

**Comment by tolleiv** [comment]
Kubernetes version: v1.21.3
PGO: v5.0.5
Postgres 13.6
Storage: Quobyte

**Comment by chobostar** [comment]
We are currently experiencing the same issue.

> b) Is there any resiliency during control-plane upgrades/downtime? "just keep on going"?
@jansmets 
1. You can put patroni to the pause mod - https://patroni.readthedocs.io/en/latest/pause.html#the-implementation
>When Patroni runs in a paused mode, it does not change the state of PostgreSQL

2. You can tune that parameters:
```
loop_wait: the number of seconds the loop will sleep. Default value: 10

ttl: the TTL to acquire the leader lock (in seconds). Think of it as the length of time before initiation of the automatic failover process. Default value: 30

retry_timeout: timeout for DCS and PostgreSQL operation retries (in seconds). DCS or network issues shorter than this will not cause Patroni to demote the leader. Default value: 10
```
to make patroni more tolerant to connectivity issues.

For example, we have that spec on our KubeAPI:
```
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: livez
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 45
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 10
    name: kube-apiserver
    ports:
    - containerPort: 6443
      hostPort: 6443
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: readyz
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 10
```
So recommended values `retry_timeout > failureThreshold + periodSeconds` or even `retry_timeout > failureThreshold + (periodSeconds + timeoutSeconds)`

Keep in my that about `ttl >= retry_timeout * 2 + loop_wait`.

I hope that information will make our cluster more stable. Good luck!

**Comment by benjaminjb** [comment]
Hello, just wanted to let y'all know if you haven't seen the attached PR, the `fsGroupChangePolicy` fix that was suggested and tested in #3023 has been merged into PGO and should be out in the next release. Reminder (for future people who find this) that this fix is only effective in K8s 1.20+

Hope this helps and if you're still experiencing a problem like this, please let the community know.
